[JSC] Add async stack traces behind the flag

sosukesuzuki opened the pull request:
#### f8a073dfbecfa286cca236c7053f3b2e72591f9d
<pre>
[JSC] Add async stack traces behind the flag
<a href="https://bugs.webkit.org/show_bug.cgi?id=296261">https://bugs.webkit.org/show_bug.cgi?id=296261</a>

Reviewed by Yusuke Suzuki.

JSC cannot obtain sufficient async stack traces. For example, this code:

    async function one(x) {
        await two(x);
    }
    async function two(x) {
        await x;
        throw new Error(&quot;error from two&quot;);
    }
    one(1).catch((err) =&gt; { print(err.stack); });

prints this stack trace:

    two@./test.js:6:20

When an error occurs within an async function, the stack trace is cut off
at the most recent await point, so higher-level functions in the async call
chain are not displayed. This happens because when an async function is
suspended at an await statement and later resumed from the microtask queue,
the original call stack is lost.

While this behavior is correct from the semantics of call stacks and microtask
queues, it is inconvenient for users.

This patch builds async stack traces through the following approach (only when
the --useAsyncStackTrace flag is enabled):

- Detect calls to `@promiseReactionJob` and `@promiseReactionJobWithoutPromise`
  during stack trace construction and retrieve the JSGenerator object hidden
  behind async-await from its arguments
- Obtain references to awaited Promises held in each generator&apos;s Context field
- Trace async context (parent generators recorded as @context fields) from the
  Promise&apos;s reaction chain
- Walk up the chain of parent generators and construct stack frames from function
  information stored in each async function&apos;s Next field

After this patch, the previous code outputs this stack trace:

    two@./WebKitBuild/Debug/test.js:6:20
    one@./WebKitBuild/Debug/test.js:2:14

V8 constructs async stack traces using a similar approach [1][2].

[1]: <a href="https://docs.google.com/document/d/13Sy_kBIJGP0XT34V1CV3nkWya4TwYx9L3Yv45LdGB6Q/edit?tab=t.0#heading=h.9ss45aibqpw2">https://docs.google.com/document/d/13Sy_kBIJGP0XT34V1CV3nkWya4TwYx9L3Yv45LdGB6Q/edit?tab=t.0#heading=h.9ss45aibqpw2</a>
[2]: <a href="https://issues.chromium.org/issues/42210758">https://issues.chromium.org/issues/42210758</a>

* JSTests/stress/async-stack-trace-basic.js: Added.
* JSTests/stress/async-stack-trace-bound-functions-basic.js: Added.
* JSTests/stress/async-stack-trace-nested-deep.js: Added.
* JSTests/stress/async-stack-trace-nested.js: Added.
* JSTests/stress/async-stack-trace-switch.js: Added.
* JSTests/stress/never-inlined-functions.js: Added.
* Source/JavaScriptCore/builtins/PromiseOperations.js:
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
* Source/JavaScriptCore/interpreter/Interpreter.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/StackFrame.cpp:
* Source/JavaScriptCore/runtime/StackFrame.h:
* Source/JavaScriptCore/tools/JSDollarVM.cpp:

Canonical link: <a href="https://commits.webkit.org/299482@main">https://commits.webkit.org/299482@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0d9f2f0f338e4ec198c0c79731e73a8c3280a428

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119146 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38827 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29478 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125365 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71209 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121024 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39523 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47409 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90496 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59879 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122099 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31506 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106804 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70919 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/30555 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24918 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69016 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/111264 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100952 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25102 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128387 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/117661 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46053 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34799 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99077 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46420 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103020 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98848 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25131 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44317 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22316 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/42632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45923 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51603 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/146359 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45390 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37640 "Found 1 new JSC binary failure: testapi, Found 18657 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_sort2.js.default, ChakraCore.yaml/ChakraCore/test/Array/push2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/ControlFlow/forInPrimitive.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48736 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47078 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -536,6 +636,15 @@ void Interpreter::getStackTrace(JSCell* owner, Vector<StackFrame>& results, size
         return IterationStatus::Continue;
     });
+
+    if (Options::useAsyncStackTrace()) [[unlikely]] {
```
If we enable it soon, we should not put unlikely.

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -516,6 +597,25 @@ void Interpreter::getStackTrace(JSCell* owner, Vector<StackFrame>& results, size
             return IterationStatus::Continue;
         }
 
+        if (Options::useAsyncStackTrace()) [[unlikely]] {
```
Dirto

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -451,14 +453,91 @@ bool Interpreter::isOpcode(Opcode opcode)
 
+    // Cache asyncFunctionResumeFunction before AssertNoGC to avoid GC
+    JSFunction* asyncFunctionResumeFunction = nullptr;
+    if (Options::useAsyncStackTrace()) [[unlikely]]
```
ditto

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -516,6 +597,25 @@ void Interpreter::getStackTrace(JSCell* owner, Vector<StackFrame>& results, size
+                if (JSFunction* callee = jsDynamicCast<JSFunction*>(visitor->callee().asCell())) {
+                    if (callee == asyncFunctionResumeFunction && visitor->argumentCountIncludingThis() > 1) {
+                        // The JSGenerator argument position varies based on optimization.
+                        // Search through all arguments until we find a JSGenerator.
```
That doesn't sound right. I think this is masking a bug. Can you investigate what is happening?

sosukesuzuki replied:
I made the fix and wrote my understanding in the comments.
https://github.com/WebKit/WebKit/pull/48314/files#diff-d32bbb01cc247b775900766b59d5d15e4db4ec21327c797d19d6dd66a7e7557dR588-R603

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -451,14 +453,91 @@ bool Interpreter::isOpcode(Opcode opcode)
+                } else {
+                    // If CodeBlock doesn't exist, create a new one
+                    CodeBlock* newCodeBlock;
+                    executable->prepareForExecution<FunctionExecutable>(vm, asyncFunction, asyncFunction->scope(), CodeSpecializationKind::CodeForCall, newCodeBlock);
```
We cannot allocate cells since these functions can be called from GC end phase.

sosukesuzuki replied:
I changed it to display only the filename without showing line numbers if the codeBlock doesn't already exist.
https://github.com/WebKit/WebKit/pull/48314/files#diff-d32bbb01cc247b775900766b59d5d15e4db4ec21327c797d19d6dd66a7e7557dR504-R509

justinmichaud commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -451,14 +453,91 @@ bool Interpreter::isOpcode(Opcode opcode)
+
+void Interpreter::getAsyncStackTrace(JSCell* owner, Vector<StackFrame>& results, JSGenerator* generator, size_t maxStackSize)
 {
+    ASSERT(Options::useAsyncStackTrace());
```
release?

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -516,6 +580,37 @@ void Interpreter::getStackTrace(JSCell* owner, Vector<StackFrame>& results, size
+                        // When DFG optimization inlines calls to @asyncFunctionResume, the JSGenerator argument
+                        // position varies depending on the caller function defined in PromiseOperations.js and
+                        // AsyncFunctionPrototype.js. DFGByteCodeParser performs inlining via inlineCall(), and
+                        // DFGStackLayoutPhase remaps argument positions using stackOffset accumulation.
```
But this is not guaranteed right? If the caller is also inlining this function too, then these arguments are unrelated.

sosukesuzuki replied:
Thank you. I've changed the approach. I've modified it to retrieve the generator from the arguments when the callee is `@promiseReactionJob` or `@promiseReactionJobWithoutPromise`, instead of `@asyncFunctionResume`. These functions are the first JavaScript functions called after resuming from `await x`. Therefore, if my understanding is correct, these functions should not be affected by inlining and should always be able to receive the generator from their arguments.

Constellation reviewed (commented):
Commented

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -451,6 +452,66 @@ bool Interpreter::isOpcode(Opcode opcode)
+        if (stateValue.isInt32()) {
+            int32_t state = stateValue.asInt32();
+            if (state > 0 && codeBlock->numberOfUnlinkedSwitchJumpTables() > 0) {
+                const UnlinkedSimpleJumpTable& jumpTable = codeBlock->unlinkedSwitchJumpTable(0);
```
I think this is the last one, correct? Since Generatorification inserts a switch after all bytecode generation gets done. Can you add a test which has integer switch inside generator to test this case?

sosukesuzuki replied:
You are right! I fixed it and added tests ( https://github.com/WebKit/WebKit/pull/48314/files#diff-a6a5bb08b318f5630df2fc8c596fbfa9f209609d46e6ebb4b63a031ece1a3cb9 )

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -451,6 +452,66 @@ bool Interpreter::isOpcode(Opcode opcode)
+                // The @context field is always at offset 3 in Promise reactions objects
+                // created by pushNewPromiseReaction in PromiseOperations.js
+                constexpr PropertyOffset contextPropertyOffset = 3;
+                JSValue contextFieldValue = reactions->getDirect(contextPropertyOffset);
```
I think this is also not guaranteed because how object gets layout can be even changed via object allocation profile. I think we need a specific object with internal fields instead of using the normal object for `ReactionsOrResult`.

sosukesuzuki replied:
It seemed like could get values using property names while avoiding GC by using `getConcurrently` and `getDirect`, so I tried implementing it that way. I think we can exclude implementing PromiseReaction using `JSInternalFieldObjectImpl` (for perf) from the scope of this PR.

```cpp
  Structure* structure = reactions->structure();
  unsigned attributes;
  PropertyOffset offset = structure->getConcurrently(vm.propertyNames->builtinNames().contextPrivateName().impl(), attributes);
  // and
  JSValue contextFieldValue = reactions->getDirect(offset);
```

I apologize if my understanding is incorrect.

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -448,6 +450,70 @@ bool Interpreter::isOpcode(Opcode opcode)
+        if (asyncFunction && !asyncFunction->isHostOrBuiltinFunction()) {
+            if (FunctionExecutable* executable = asyncFunction->jsExecutable()) {
+                // If a CodeBlock doesn't already exist, the stack trace will only show the filename and won't show line column
+                if (CodeBlock* codeBlock = executable->eitherCodeBlock()) {
```
eitherCodeBlock is wrong since construct / call have completely different CodeBlock. Definitely one of them should be picked.

sosukesuzuki replied:
I changed it to pick the call here

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -513,6 +581,26 @@ void Interpreter::getStackTrace(JSCell* owner, Vector<StackFrame>& results, size
+                if (JSFunction* callee = jsDynamicCast<JSFunction*>(visitor->callee().asCell())) {
+                    JSGlobalObject* globalObject = callFrame->lexicalGlobalObject(vm);
+
+                    JSGenerator* generator = nullptr;
+                    if (callee == globalObject->promiseReactionJobFunction())
+                        generator = jsDynamicCast<JSGenerator*>(visitor->callFrame()->argument(3));
+                    else if (callee == globalObject->promiseReactionJobWithoutPromiseFunction())
+                        generator = jsDynamicCast<JSGenerator*>(visitor->callFrame()->argument(2));
```
Yeah, this is not correct. StackVisitor can populate each frame's callee. And there is no guarantee whether each function gets inlined or not. If these functions are inlined, then CallFrame is the inliner's CallFrame, and arguments are their arguments. You need to have some mechanism to ensure they are not inlined, or implementing them in C++, or something. Currently, it can be wrong, based on inlining heuristics.

sosukesuzuki replied:
I changed it to prevent inlining using the `@neverInline` attribute added in https://github.com/WebKit/WebKit/pull/49873

sosukesuzuki committed:
[JSC] Add async stack traces behind the flag
https://bugs.webkit.org/show_bug.cgi?id=296261

Reviewed by Yusuke Suzuki.

JSC cannot obtain sufficient async stack traces. For example, this code:

    async function one(x) {
        await two(x);
    }
    async function two(x) {
        await x;
        throw new Error("error from two");
    }
    one(1).catch((err) => { print(err.stack); });

prints this stack trace:

    two@./test.js:6:20

When an error occurs within an async function, the stack trace is cut off
at the most recent await point, so higher-level functions in the async call
chain are not displayed. This happens because when an async function is
suspended at an await statement and later resumed from the microtask queue,
the original call stack is lost.

While this behavior is correct from the semantics of call stacks and microtask
queues, it is inconvenient for users.

This patch builds async stack traces through the following approach (only when
the --useAsyncStackTrace flag is enabled):

- Detect calls to `@promiseReactionJob` and `@promiseReactionJobWithoutPromise`
  during stack trace construction and retrieve the JSGenerator object hidden
  behind async-await from its arguments
- Obtain references to awaited Promises held in each generator's Context field
- Trace async context (parent generators recorded as @context fields) from the
  Promise's reaction chain
- Walk up the chain of parent generators and construct stack frames from function
  information stored in each async function's Next field

After this patch, the previous code outputs this stack trace:

    two@./WebKitBuild/Debug/test.js:6:20
    one@./WebKitBuild/Debug/test.js:2:14

V8 constructs async stack traces using a similar approach [1][2].

[1]: https://docs.google.com/document/d/13Sy_kBIJGP0XT34V1CV3nkWya4TwYx9L3Yv45LdGB6Q/edit?tab=t.0#heading=h.9ss45aibqpw2
[2]: https://issues.chromium.org/issues/42210758

* JSTests/stress/async-stack-trace-basic.js: Added.
* JSTests/stress/async-stack-trace-bound-functions-basic.js: Added.
* JSTests/stress/async-stack-trace-nested-deep.js: Added.
* JSTests/stress/async-stack-trace-nested.js: Added.
* JSTests/stress/async-stack-trace-switch.js: Added.
* JSTests/stress/never-inlined-functions.js: Added.
* Source/JavaScriptCore/builtins/PromiseOperations.js:
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
* Source/JavaScriptCore/interpreter/Interpreter.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/StackFrame.cpp:
* Source/JavaScriptCore/runtime/StackFrame.h:
* Source/JavaScriptCore/tools/JSDollarVM.cpp:

Canonical link: https://commits.webkit.org/299482@main
